File: src/backend/src/schemas/risk.py
"""
Pydantic schemas for risk assessment data validation and serialization in the Customer Success AI Platform.
Provides comprehensive validation rules and performance optimizations for risk profiles and assessments.

Version: pydantic 2.x
"""

from datetime import datetime
from typing import Dict, List, Optional
from uuid import UUID
from pydantic import BaseModel, Field, field_validator, model_config

from models.risk import RISK_SEVERITY_LEVELS, RISK_SCORE_THRESHOLDS

# Risk factor weight configuration
RISK_FACTOR_WEIGHTS = {
    "usage_decline": 0.35,
    "support_tickets": 0.25,
    "payment_delays": 0.25,
    "engagement_score": 0.15
}

class RiskFactorBase(BaseModel):
    """Base schema for risk factor data validation with enhanced validation rules."""
    
    model_config = model_config(from_attributes=True, validate_assignment=True)

    impact_score: float = Field(
        ...,
        ge=0.0,
        le=1.0,
        description="Impact score of the risk factor (0-1)"
    )
    category: str = Field(
        ...,
        min_length=1,
        max_length=50,
        description="Category of the risk factor"
    )
    description: str = Field(
        ...,
        min_length=1,
        max_length=500,
        description="Detailed description of the risk factor"
    )
    metadata: Dict = Field(
        default_factory=dict,
        description="Additional metadata for the risk factor"
    )
    confidence_score: float = Field(
        default=1.0,
        ge=0.0,
        le=1.0,
        description="Confidence level in the risk factor assessment"
    )
    assessed_at: datetime = Field(
        default_factory=datetime.utcnow,
        description="Timestamp of risk factor assessment"
    )

    @field_validator('impact_score')
    def validate_impact_score(cls, value: float) -> float:
        """Validates impact score is within acceptable range."""
        return round(max(0.0, min(1.0, value)), 3)

class RiskProfileCreate(BaseModel):
    """Schema for creating new risk profile entries with comprehensive validation."""
    
    model_config = model_config(from_attributes=True, validate_assignment=True)

    customer_id: UUID = Field(
        ...,
        description="UUID of the associated customer"
    )
    score: float = Field(
        ...,
        ge=0.0,
        le=100.0,
        description="Overall risk score (0-100)"
    )
    factors: List[RiskFactorBase] = Field(
        ...,
        min_items=1,
        description="List of risk factors contributing to the score"
    )
    recommendations: Dict = Field(
        default_factory=dict,
        description="Automated recommendations based on risk assessment"
    )
    version: str = Field(
        default="1.0",
        pattern=r"^\d+\.\d+$",
        description="Schema version for compatibility tracking"
    )
    metadata: Dict = Field(
        default_factory=dict,
        description="Additional risk profile metadata"
    )

    @field_validator('score')
    def validate_score(cls, value: float) -> float:
        """Validates risk score with enhanced rules."""
        if value >= RISK_SCORE_THRESHOLDS['CRITICAL']:
            return round(min(100.0, value), 2)
        return round(max(0.0, value), 2)

class RiskProfileResponse(BaseModel):
    """Schema for risk profile API responses with computed fields."""
    
    model_config = model_config(from_attributes=True, validate_assignment=True)

    id: UUID = Field(..., description="Unique identifier for the risk profile")
    customer_id: UUID = Field(..., description="Associated customer identifier")
    score: float = Field(..., description="Overall risk score")
    severity_level: int = Field(..., description="Computed severity level")
    factors: List[RiskFactorBase] = Field(..., description="Risk factors")
    recommendations: Dict = Field(..., description="Intervention recommendations")
    assessed_at: datetime = Field(..., description="Assessment timestamp")
    created_at: datetime = Field(..., description="Creation timestamp")
    updated_at: datetime = Field(..., description="Last update timestamp")
    version: str = Field(..., description="Schema version")
    metadata: Dict = Field(..., description="Additional metadata")
    confidence_score: float = Field(
        default=1.0,
        description="Overall confidence in risk assessment"
    )
    trend_analysis: Dict = Field(
        default_factory=dict,
        description="Historical trend analysis"
    )

    @property
    def compute_severity(self) -> int:
        """Computes severity level from risk score."""
        if self.score >= RISK_SCORE_THRESHOLDS['CRITICAL']:
            return RISK_SEVERITY_LEVELS['CRITICAL']
        elif self.score >= RISK_SCORE_THRESHOLDS['HIGH']:
            return RISK_SEVERITY_LEVELS['HIGH']
        elif self.score >= RISK_SCORE_THRESHOLDS['MEDIUM']:
            return RISK_SEVERITY_LEVELS['MEDIUM']
        return RISK_SEVERITY_LEVELS['LOW']

class RiskProfileUpdate(BaseModel):
    """Schema for updating existing risk profiles with validation."""
    
    model_config = model_config(from_attributes=True, validate_assignment=True)

    score: Optional[float] = Field(
        None,
        ge=0.0,
        le=100.0,
        description="Updated risk score"
    )
    factors: Optional[List[RiskFactorBase]] = Field(
        None,
        min_items=1,
        description="Updated risk factors"
    )
    recommendations: Optional[Dict] = Field(
        None,
        description="Updated recommendations"
    )
    version: Optional[str] = Field(
        None,
        pattern=r"^\d+\.\d+$",
        description="Schema version"
    )
    metadata: Optional[Dict] = Field(
        None,
        description="Updated metadata"
    )

File: src/web/src/types/risk.ts
/**
 * @fileoverview TypeScript type definitions for risk assessment and scoring functionality
 * Defines types and interfaces for the Customer Success AI Platform's risk assessment capabilities
 */

/**
 * Enumeration of possible risk severity levels
 */
export enum RiskLevel {
    LOW = 'LOW',
    MEDIUM = 'MEDIUM',
    HIGH = 'HIGH',
    CRITICAL = 'CRITICAL'
}

/**
 * Type representing valid risk score range (0-100)
 */
export type RiskScoreRange = number;

/**
 * Union type of possible risk factor categories
 */
export type RiskFactorCategory = 'usage' | 'support' | 'payment' | 'engagement';

/**
 * Interface defining risk factor weight configuration
 */
export interface RiskFactorWeight {
    usageDecline: number;    // Weight: 0.35
    supportTickets: number;  // Weight: 0.25
    paymentDelays: number;   // Weight: 0.25
    engagementScore: number; // Weight: 0.15
}

/**
 * Interface for individual risk factors contributing to overall risk assessment
 */
export interface RiskFactor {
    category: RiskFactorCategory;
    impactScore: RiskScoreRange;
    description: string;
    metadata: Record<string, unknown>;
}

/**
 * Interface for risk score with associated factors and recommendations
 */
export interface RiskScore {
    score: RiskScoreRange;
    level: RiskLevel;
    factors: readonly RiskFactor[];
    recommendations: Record<string, unknown>;
}

/**
 * Interface for complete risk assessment data structure
 */
export interface RiskAssessment {
    id: string;
    customerId: string;
    score: RiskScoreRange;
    severityLevel: RiskLevel;
    factors: readonly RiskFactor[];
    recommendations: Record<string, unknown>;
    assessedAt: Date;
    createdAt: Date;
    updatedAt: Date;
}

/**
 * Risk score threshold constants for determining risk levels
 */
export const RISK_SCORE_THRESHOLDS = {
    LOW: 25,
    MEDIUM: 50,
    HIGH: 75,
    CRITICAL: 90
} as const;

/**
 * Risk factor weight constants for score calculation
 */
export const RISK_FACTOR_WEIGHTS: Readonly<RiskFactorWeight> = {
    usageDecline: 0.35,
    supportTickets: 0.25,
    paymentDelays: 0.25,
    engagementScore: 0.15
} as const;